Add distributed log configuration system

benclifford opened the pull request:
## Summary

This PR introduces the concept of distributed Python `logging` configuration:

i) a user might want the same log configuration in several processes, because their workflow comprises several processes that should be observed together - an example would be a laptop-based workflow orchestration/submit side, and an HPC worker node based execution process.

ii) several authorities might regard themselves as the one entitled to configure logging for a Python process. An example here might be academy launching an agent in a remote Parsl worker inside a Globus Compute endpoint (and wanting to observe log information), and the enclosing Globus Compute endpoint also wanting to observe log information.

iii) users, implementers and researchers are interested in observing logging novel means: for example, under development distributed logging systems -- see diaspora or via systems focused on specific uses (for example) provenance - see flowcept.

These are handled quite badly/informally at the moment: Academy does some remote log configuration, but using logger.basicConfiguration which only happens if no one else has configured the root logger. That means it is very sensitive to what anyone else has done. The academy remote configuration is also restricted to a tiny language which can describe file targets on the remote file system and nothing else.

The notion of log configuration introduced in this PR is:

Log configurations are described by `LogConfig` objects, each of a class that knows how to configure logging in a particular way.

These configuration objects are expected to be serialized to other Python processes, primarily using pickle. In the presence of multiple serializations, these objects can still tell if they are from the same original source object (by use of an identifier field, rather than expecting the serialization system to preserve this identity relation).

These configuration objects are expected to know how to initialize their local process in their own particular way, on demand, and to know how to undo that configuration on demand.

None of these log configuration objects should expect to be the main log configuration for a process and should always expect to be alongside other configurations (either of this form, or set up by some other piece of Python code) - they are *visiting* a process in the sense that a Parsl task is a visitor inside a Parsl worker process , which has a separate existence (and so an academy worker inside a Parsl task is also a visitor).

Some academy scaffolding (`log_context`) manages these log configurations - for example as a context manager - to avoid duplicate initializations or too-early de-initializations.

This PR splits the log configuration from before this PR into a couple of different `LogConfig` implementations:

`ConsoleLogging` - which knows about logging to stdout, including with ANSI color.

`FileLogging` - which knows about logging to a file on the co-located filesystem.

There is a new `LogConfig` called `JSONPool` which knows how to log to an automatically named directory under the users home directory, in JSON format. This is intended for machine consumption.

There is a new `MultiLogConfig` which joins several `LogConfig` objects into one. This supports use cases such as logging to a file and also to some other experimental system simultanously.

Outside of this PR, I have also prototyped a `LogConfig` which knows how to send some logs into flowcept - https://github.com/ORNL/flowcept. I have also worked with Diaspora Octopus to drive some ideas for implementation, though I don't have a working implementation right now.

## Related Issues

## Changes

- [x] Breaking (backwards incompatible changes to public interfaces)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (non-breaking change or feature addition)
- [ ] Refactor (internal code or design clean up)
- [ ] Documentation (no changes to the code)
- [ ] Test (changes or additions to testing)
- [ ] Build (change to CI workflows or build processes)
- [ ] Package (changes to package metadata or dependency versions)

## Testing

Added a bunch of tests around logging to the test suite.

## Pull Request Checklist

Please confirm the PR meets the following requirements.
- [x] Relevant tags are added based on the types of changes.
- [x] Code changes pass `pre-commit` (e.g., ruff, mypy, etc.).
- [x] Tests have been added to show the fix is effective or that the new feature works.
- [x] New and existing unit tests pass locally with the changes.
- [x] Docs have been updated and reviewed if relevant.

AK2000 commented on academy/manager.py:
```diff
@@ -481,17 +477,25 @@ async def launch(  # noqa: PLR0913
+        #        f'Manager. `init_logging` should only be called once per '
+        #        f'process.',
+        #        stacklevel=2,
+        #    )
```
This was to avoid what I thought might be a common pitfall where people set `init_logging=true` when they didn't need to. `init_logging` multiple times per htex worker (or per process pool worker) is not much of an issue because they can override each other (i.e. the new agent running in that worker supplants anything else running in that worker, so the old thing doesn't log any more).

benclifford replied:
The (mis)behaviour that this warning is about should be gone now -- there can be multiple active/overlapping configs, and the tests in this PR test that.

AK2000 reviewed (commented):
This looks like a good direction to me. How do you picture this integrating with some distributed logging solution (that we talked about building upon rather than reimplementing). Were you thinking that those plug in as another `ObservabilityConfig`?

benclifford commented:
> This looks like a good direction to me. How do you picture this integrating with some distributed logging solution (that we talked about building upon rather than reimplementing). Were you thinking that those plug in as another `ObservabilityConfig`?

yeah -- @ryanchard and I have prototyped this a little bit as "just another Handler" that also needs some context/configuration about connecting to however the distributed logging system works -- eg server endpoint, credentials, whatever is relevant for that particular method.

benclifford commented on academy/observability/examples.py:
```diff
@@ -0,0 +1,122 @@
+    systems under the same pool instance ID.
+
+    The logs are not configurable: they are full debug logs from the root,
+    in JSON format. That's intended to get *rid* of configurability for users
```
i'm fine adding in configurability, it was mostly the lazy implementation option. mostly I was pushing towards the style of log everything and if you think you're such an advanced user you can turn this off, then we'll let you do that.

benclifford committed:
Introduce a massive rearrangement of logging configuration

This is intended to make distributed log configuration a serious feature
of Academy, rather than somebody else's problem.

This PR moves log configuration from sort-of-once-per-process configuration
to context scoped regions which may overlap in one process and may be
share configurations across multiple processes.

For a change-oriented overview, see this PR's modification of
docs/migrations.md

For non-change description, start at this PR's docstring for
academy.logging.configs.base.LogConfig

I expect substantial changes later on as/when people make use of this.


One quote from alok about the old system that this addresses.

# This comes from a quote from Alok in github
# https://github.com/academy-agents/academy/pull/295#discussion_r2604534531
# that I am not satisfied is true and would like tested:
# init_logging multiple times per htex worker (or per process pool worker)
# is not much of an issue because they can override each other (i.e. the
# new agent running in that worker supplants anything else running in that
# worker, so the old thing doesn't log any more).

benclifford commented on academy/exchange/cloud/app.py:
```diff
@@ -486,9 +487,19 @@ def create_app(
+
+    # TODO: in the parsl prototype, I pass in a bunch of context here
+    # and I wonder if that's sensible here too? parsl has at least the
+    # notion of a rundir, which academy does not, and also has named
```
I removed changes to cloud/app from this PR -- that application has enough of its own configuration system that I am leaving that to a separate PR.

benclifford commented on tests/integration/sequential_agent_logs_test.py:
```diff
@@ -0,0 +1,111 @@
+            await handle.shutdown()
+            logger.info('after A shutdown')
+
+            # does this returning ^ mean that all shutdown activity (eg log
```
I used manager.wait (having forgotten about it twice...)

AK2000 commented on examples/01-actor-client/run-01.py:
```diff
@@ -27,21 +27,21 @@ async def get_count(self) -> int:
 async def main() -> int:
-    init_logging(logging.INFO)
+    lc = recommended_logging()
+    with log_context(lc):
```
I see this being an extremely common use case/pattern (i.e. a Manager nested in a log context, where the log config is then passed to every agent), but I don't really like the feel of having essentially the whole program nested within 2 context managers. What do you think about the manager accepting a log config, entering/exiting the log config as part of it's own context, and using the log config it was initialized with as a default to pass to all agents?

benclifford replied:
yeah I think that tidies things up a lot, syntax-wise - I'll do that in this PR because it will also affect all the examples.

benclifford replied:
I've changed this. It looks much nicer!

AK2000 commented on academy/logging/configs/jsonpool.py:
```diff
@@ -0,0 +1,70 @@
+        # can be different on every invocation as the config object is moved
+        # between execution hosts.
+        path = (
+            pathlib.Path.home() / '.academy' / 'logs' / self.uuid / instance_id
```
I see creating this being difficult to parse --- the uuid of the entire log config is random and so is the instance id. When trying to analyze runs, how do I unpack which run and which instance I should look at? Could we create this folder (or maybe the entire log config?) with a date-time id (e.g. 20260514104200-uuid[:8]) or something like that to make distinguishing it easier? I guess my high level goal would be to easily associate directories with a specific run (and for a specific agent if we are creating different directories for each agent as I think would happen in this case).

benclifford replied:
haha I was (at least when prototyping this) going the other way and *not* putting identities in the filenames, deliberately trying to model having eg a giant cloud hosted log pool that just stores a pile of log records and you have to figure this stuff out from the records themselves. which itself is probably influenced by not having a strong concept of a "run" when there's all different things with different lifetimes everywhere. and in that sense, you (a human) aren't meant to parse this - it's only to chop up the giant pool of records to avoid conflicting file names.

it should be pretty easy to timestamp/label that top directory so that a "run" is (e.g.) a creation of a log config (which is then passed to a manager and propagates, according to the other comment you made).

Naming the instance ID is something more complicated because log contexts don't have names as such -- but in the parsl version of this work, I have encountered something similar, so I think I'll have to figure this out. But I don't think it should block this #295 PR -- the jsonpool is in here as a demo to further develop rather than something I am treating as a finished feature.

AK2000 commented on academy/logging/configs/jsonpool.py:
```diff
@@ -0,0 +1,70 @@
+        # can be different on every invocation as the config object is moved
+        # between execution hosts.
+        path = (
+            pathlib.Path.home() / '.academy' / 'logs' / self.uuid / instance_id
```
Can we also unify this path with the default token store path (which right now is ~/local/share/academy). I don't know which one should be preferred, but I don't like having two different meta-data directories for academy.

benclifford replied:
yep - I'll use the path that is already used in the existing `main` branch.

AK2000 commented on academy/logging/recommended.py:
```diff
@@ -0,0 +1,39 @@
+    logfile: str | pathlib.Path | None = None,
+    logfile_level: int | str | None = None,
+    color: bool = True,
+    extra: int = False,
```
I know you took this interface from the current `init_logging` but personally I don't want the `extra` logs to always come with the thread information (it doesn't hurt, but I haven't gotten a lot of use out of thread information). If we are re-doing this interface anyway, what are your thoughts on breaking these into different options?

benclifford replied:
I think its fine to break up the extra= parameter into specific feature flags.

benclifford replied:
i'll do this in a subsequent PR

AK2000 reviewed (dismissed):
Looks good to me. Minor comments (I think most if not all could be left to a separate PR if you want)
